lib/services/plugins/npm-plugins-service-base.ts

```diff
@@ -0,0 +1,290 @@
+			let pluginUrl = plugin && plugin.data && plugin.data.Url ? plugin.data.Url : null;
+			let plugins = this.findPlugins([pluginIdentifier]).wait();
+			let pluginKeys = _.map(plugins, (pluginInfo: IBasicPluginInformation) => pluginInfo.name);
+			let pluginsCount = pluginKeys.length;
+
+			if (pluginsCount === 0) {
+				if (pluginUrl) {
+					try {
+						// The plugin is not in npm but it is in our marketplace.
+						this.fetchPluginCore(pluginUrl).wait();
+					} catch (error) {
+						this.$logger.info("The plugin cannot be downloaded using npm, because it has no package.json in it. You can still download it from this link: " + plugin.data.Url.grey);
+					}
+				} else {
+					this.fetchPluginCore(pluginIdentifier).wait();
+				}
+
+				return;
+			}
+
+			if (pluginsCount > 1 && pluginKeys[0] !== pluginIdentifier) {
+				if (commonHelpers.isInteractive()) {
+					let selectedPlugin = this.$prompter.promptForChoice("We found multiple plugins with your search parameters please choose which one you want to fetch.", pluginKeys).wait();
+					this.fetchPluginCore(selectedPlugin).wait();
+				} else {
+					this.$logger.out("There are more then 1 matching plugins: " + pluginKeys.join(", ") + ".");
+				}
+
+				return;
+			}
+
+			try {
+				this.fetchPluginCore(pluginKeys[0]).wait();
+			} catch (err) {
+				if (pluginUrl) {
+					this.$logger.trace("Error while trying to fetch plugin with id " + pluginIdentifier + " via npm. Error is: " + err.message + ".");
+					this.fetchPluginCore(pluginUrl).wait();
+				} else {
+					this.$errors.fail(err.message);
+				}
+			}
+		}).future<void>()();
+	}
+
+	public abstract getAvailablePlugins(pluginsCount?: number): IPlugin[];
+
+	public abstract getInstalledPlugins(): IPlugin[];
+
+	public abstract printPlugins(plugins: IPlugin[]): void;
+
+	public abstract addPlugin(pluginIdentifier: string): IFuture<void>;
+
+	public abstract removePlugin(pluginName: string): IFuture<void>;
+
+	public abstract configurePlugin(pluginName: string, version?: string, configurations?: string[]): IFuture<void>;
+
+	public abstract isPluginInstalled(pluginName: string): boolean;
+
+	public abstract getPluginBasicInformation(pluginName: string): IFuture<IBasicPluginInformation>;
+
+	public abstract filterPlugins(plugins: IPlugin[]): IFuture<IPlugin[]>;
+
+	protected isPluginFetched(pluginName: string): IFuture<boolean> {
+		return ((): boolean => {
+			// Fetched plugins are in the "plugins" directory both for Cordova and NativeScript projects.
+			let projectPluginsDirectory = path.join(this.$project.projectDir, "plugins");
+			let filterOptions = { enumerateDirectories: true, includeEmptyDirectories: false };
+			let fetchedPlugins = this.$fs.enumerateFilesInDirectorySync(projectPluginsDirectory, (item: string) => {
+				let itemBaseName = path.basename(item);
+				return this.hasTgzExtension(itemBaseName) ? itemBaseName.indexOf(pluginName) >= 0 : itemBaseName === pluginName;
```

Comment:
this seems incorrect - can't we just remove the `.tgz` extension and use strong equality check instead of "indexOf":
```TypeScript
return itemBaseName.replace(".tgz", "") === pluginName;
```